packages/itwin/tree-widget/src/tree-widget-react/components/trees/classifications-tree/ClassificationsTreeDefinition.ts

```diff
@@ -330,3 +353,400 @@ function createClassificationHasChildrenSelector(classificationAlias: string) {
+      ];
+      const ecsql = `
+        SELECT * FROM (
+          SELECT
+            ct.ClassName AS ClassName,
+            ct.ECInstanceId AS ECInstanceId
+          FROM
+            ${CLASSIFICATION_TABLES_WITH_LABELS_CTE} ct
+          WHERE
+            ct.DisplayLabel LIKE '%' || ? || '%' ESCAPE '\\'
+
+          ${
+            classificationIds.length > 0
+              ? `
+                UNION ALL
+
+                SELECT
+                  c.ClassName AS ClassName,
+                  c.ECInstanceId AS ECInstanceId
+                FROM
+                  ${CLASSIFICATIONS_WITH_LABELS_CTE} c
+                WHERE
+                  c.DisplayLabel LIKE '%' || ? || '%' ESCAPE '\\'
+
+                UNION ALL
+
+                SELECT
+                  e2d.ClassName AS ClassName,
+                  e2d.ECInstanceId AS ECInstanceId
+                FROM
+                  ${ELEMENTS_2D_WITH_LABELS_CTE} e2d
+                WHERE
+                  e2d.DisplayLabel LIKE '%' || ? || '%' ESCAPE '\\'
+
+                UNION ALL
+
+                SELECT
+                  e3d.ClassName AS ClassName,
+                  e3d.ECInstanceId AS ECInstanceId
+                FROM
+                  ${ELEMENTS_3D_WITH_LABELS_CTE} e3d
+                WHERE
+                  e3d.DisplayLabel LIKE '%' || ? || '%' ESCAPE '\\'
+              `
+              : ""
+          }
+        )
+        ${props.limit === undefined ? `LIMIT ${MAX_FILTERING_INSTANCE_KEY_COUNT + 1}` : props.limit !== "unbounded" ? `LIMIT ${props.limit}` : ""}
+      `;
+      const bindings = [
+        { type: "string" as const, value: adjustedLabel },
+        ...(classificationIds.length > 0
+          ? [
+              { type: "string" as const, value: adjustedLabel },
+              { type: "string" as const, value: adjustedLabel },
+              { type: "string" as const, value: adjustedLabel },
+            ]
+          : []),
+      ];
+      return { ctes, ecsql, bindings };
+    }).pipe(
+      mergeMap((queryProps) => {
+        if (!queryProps) {
+          return EMPTY;
+        }
+        return props.imodelAccess.createQueryReader(queryProps, { restartToken: "tree-widget/classifications-tree/filter-by-label-query", limit: props.limit });
+      }),
+      map((row): InstanceKey => {
+        let className: string;
+        switch (row.ClassName) {
+          case "ct":
+            className = CLASS_NAME_ClassificationTable;
+            break;
+          case "c":
+            className = CLASS_NAME_Classification;
+            break;
+          case "e2d":
+            className = CLASS_NAME_GeometricElement2d;
+            break;
+          default:
+            className = CLASS_NAME_GeometricElement3d;
+            break;
+        }
+        return {
+          className,
+          id: row.ECInstanceId,
+        };
+      }),
+      toArray(),
+      mergeMap((targetItems) => createInstanceKeyPathsFromTargetItems({ ...props, targetItems })),
```

Comment:
Could we avoid creating an array of all instance keys? It seems we only need that to validate the count, but I think `createInstanceKeyPathsFromTargetItems` could accumulate the number of keys it receives and throw if/when the count is exceeded.